spark/src/main/scala/org/apache/spark/sql/delta/expressions/HilbertIndex.scala

```diff
@@ -0,0 +1,405 @@
+ * That represents the information necessary to construct a state list for a given number
+ * of dimension, N.
+ * Once you have the generator table for your dimension you can construct a state list.
+ * You can then turn those state lists into compact state lists that store all the information
+ * in one large array of longs.
+ */
+// scalastyle:on line.size.limit
+object HilbertIndex {
+
+  private type CompactStateList = HilbertCompactStateList
+
+  val SIZE_OF_INT = 32
+
+  /**
+   * Construct the generator table for a space of dimension n.
+   * This table consists of 2^n rows, each row containing Y, X1, X2, dY and TY.
+   *   Y    The index in the array representing the table. (0 to (2^n - 1))
+   *   X1   A coordinate representing points on the curve expressed as an n-point.
+   *        These are arranged such that if two rows differ by 1 in Y then the binary
+   *        representation of their X1 values differ by exactly one bit.
+   *        These are the "Gray-codes" of their Y value.
+   *   X2   A pair of n-points corresponding to the first and last points on the first
+   *        order curve to which X1 transforms in the construction of a second order curve.
+   *   dY   Represents the magnitude of difference between X2 values in this row.
+   *   TY   A transformation matrix that transforms X2(1) to the X1 value where Y is zero and
+   *        transforms X2(2) to the X1 value where Y is (2^n - 1)
+   */
+  def getStateGenerator(n: Int): GeneratorTable = {
+    val x2s = getX2GrayCodes(n)
+
+    val len = 1 << n
+    val rows = (0 until len).map { i =>
+      val x21 = x2s(i << 1)
+      val x22 = x2s((i << 1) + 1)
+      val dy = x21 ^ x22
+
+      Row(
+        y = i,
+        x1 = i ^ (i >>> 1),
+        x2 = (x21, x22),
+        dy = dy,
+        m = HilbertMatrix(n, x21, getSetColumn(n, dy))
+      )
+    }
+
+    new GeneratorTable(n, rows)
+  }
+
+  // scalastyle:off line.size.limit
+  /**
+   * This will construct an x2-gray-codes sequence of order n as described in
+   *  https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=bfd6d94c98627756989b0147a68b7ab1f881a0d6
+   *
+   *   Each pair of values corresponds to the first and last coordinates of points on a first
+   *   order curve to which a point taken from column X1 transforms to at the second order.
+   */
+  // scalastyle:on line.size.limit
+  private[this] def getX2GrayCodes(n: Int) : Array[Int] = {
+    if (n == 1) {
+      // hard code the base case
+      return Array(0, 1, 0, 1)
+    }
+    val mask = 1 << (n - 1)
+    val base = getX2GrayCodes(n - 1)
+    base(base.length - 1) = base(base.length - 2) + mask
+    val result = Array.fill(base.length * 2)(0)
+    base.indices.foreach { i =>
+      result(i) = base(i)
+      result(result.length - 1 - i) = base(i) ^ mask
+    }
+    result
+  }
+
+  private[this] case class Row(y: Int, x1: Int, x2: (Int, Int), dy: Int, m: HilbertMatrix)
```

Comment:
Not sure how many of these will be in memory at a time, but using a tuple for `x2` results in an extra object allocation and risk of non-locality... `Tuple2` _is_ specialized for `(Int, Int)` so the Ints themselves won't be _stored_ as boxes, but the `Tuple2` itself still needs to exist as an intermediary object.